text/0000-rtsan-and-attributes.md

```diff
@@ -0,0 +1,270 @@
+    #8 0x0001021a67b8 in std::rt::lang_start::ha4c268826019738b rt.rs:163
+    #9 0x0001021a68b4 in main+0x20 (example:arm64+0x1000028b4)
+    #10 0x00018d8860dc  (<unknown module>)
+    #11 0x4c3d7ffffffffffc  (<unknown module>)
+
+SUMMARY: RealtimeSanitizer: blocking-call (example:arm64+0x100002940) in rtsan::notify_blocking_call::ha111b6bcc1e1c566+0x24
+fish: Job 1, 'cargo run --package example --f…' terminated by signal SIGABRT (Abort)
+
+```
+
+### The addition of the `rtsan_scoped_disabler!` macro
+
+It will be important to allow users to opt-out of rtsan detection for a specific scope. This may be useful if the end user thinks RTsan has a false positive, or it happens in third-party code they don't control.
+
+We propose addition of the `rtsan_scoped_disabler!` macro:
+```rust
+#[nonblocking]
+fn process() {
+  rtsan_scoped_disabler!({
+        let audio = vec![1.0; 256]; // report is suppressed
+  });
+}
+```
+
+When rtsan is not enabled with RUSTFLAGS, this will become a no-op, so it will be safe for users to leave in their code. 
+
+We are open to advice on which file this macro belongs in.
+
+### Run-time suppression list
+There is one other way to opt-out of rtsan checking which will automatically work with Rust.
+
+Users may specify suppression lists, which may be passed in via an environment variable
+```
+> cat suppressions.supp
+call-stack-contains:*spin*
+> RTSAN_OPTIONS=suppressions=suppressions.supp cargo run
+```
+
+### `no_sanitize`
+Another approach we could take is similar to the ASan and TSan `no_sanitize` attribute. We advocate for the scoped disabler macro, as it allows users to specify a more specific scope to disable the tool in. This means users will not have to extract real-time unsafe code into helper functions to disable them at the function level.
+
+To match the other sanitizers, adding in `no_sanitize` could be considered instead of/in addition to the macro, depending on input on this RFC.
+
+### Disabling RealtimeSanitizer in the `panic!` and all `assert!` macros.
+
+If users rely on `panic!` or `assert!` while running under RealtimeSanitizer, they will hit an intercepted call before the message is printed.
+
+For example:
+
+```rust
+#[nonblocking]
+fn processor(buffer: &[f32]) {
+    buffer[512]; // Oops, out of bounds!! should panic!
+}
+
+processor(&[0.0; 512]);
+```
+
+Running under RTSan the user gets this message. This is due to some memory being allocated to prepare to print the panic message.
+```
+==31969==ERROR: RealtimeSanitizer: unsafe-library-call
+Intercepted call to real-time unsafe function `malloc` in real-time context!
+```
+
+A user should expect this out of bounds access to print:
+```
+index out of bounds: the len is 512 but the index is 512
+```
+
+To adhere to this expected behavior, RTSan should be disabled for `panic!` and each of the assertion macros: `assert`, `assert_eq`, `assert_ne`, `debug_assert`, `debug_assert_eq`, `debug_assert_ne`.
```

Comment:
Note that it would be possible to make `panic!` and the following backtrace process not allocate, but obviously it currently does.